code/compact-trie.lisp

```diff
@@ -0,0 +1,341 @@
+  '(or compact-entry vector-of-compact-entry))
+
+(defclass compact-leaf-mixin (leaf-mixin)
+  ((%entries :initarg :entries
+             :type    compact-entries)))
+
+(defmethod utilities.print-items:print-items append
+    ((object compact-leaf-mixin))
+  (let ((entries '()))
+    (map-entries (lambda (entry) (push entry entries))
+                 object  (%entries object))
+    `(((:entries (:after :children)) " entries: ~{~A~^ ~}" ,entries))))
+
+(defmethod map-entries
+    ((function function) (node compact-leaf-mixin) (entries t))
+  (funcall function entries))
+
+(defmethod map-entries
+    ((function function) (node compact-leaf-mixin) (entries vector))
+  (map nil function entries))
+
+(defmethod %lookup ((function function)
+                    (string   string)
+                    (suffix   (eql 0))
+                    (node     compact-leaf-mixin))
+  (map-entries (lambda (entry)
+                 (let ((word (expand-entry entry string)))
+                   (funcall function word)))
+               node (%entries node)))
+
+(defmethod compact-node-slots append ((node leaf-mixin) (depth integer))
+  (let ((compact-entries '()))
+    (map-entries (lambda (entry)
+                   (push (compact-entry entry) compact-entries))
+                 node (%entries node))
+    (let ((new-entries
+            (cond ((a:length= 1 compact-entries)
+                   (first compact-entries))
+                  ((every (a:of-type 'unsigned-byte) compact-entries)
+                   (let ((bit-count (reduce #'max compact-entries
+                                            :key #'integer-length)))
+                     (make-array (length compact-entries)
+                                 :element-type     `(unsigned-byte ,bit-count)
+                                 :initial-contents compact-entries)))
+                  (t
+                   (coerce compact-entries 'vector)))))
+      (list :entries new-entries))))
+
+;;; `compact-interior-mixin'
+
+(deftype child-key/character ()
+  'character)
+
+(deftype child-key/string ()
+  'simple-string)
+
+(deftype child-key ()
+  '(or child-key/character child-key/string))
+
+(deftype compact-child-cell ()
+  '(cons child-key (or null ; can happen temporarily when rebuilding from fasl
+                       compact-node
+                       compact-entries)))
+
+(defun %every-compact-child-chell (object)
+  (and (typep object 'sequence)
+       (let ((length (length object)))
+         (and (zerop (mod length 2))
+              (loop :for i :below length :by 2
+                    :for key   = (aref object (+ i 0))
+                    :for child = (aref object (+ i 1))
+                    :always (and (typep key 'child-key)
+                                 (typep child '(or null ; can happen temporarily while rebuilding from fasl
+                                                   compact-node
+                                                   compact-entries))))))))
```

Comment:
Caught it upon re-reading - the `(or ...)` is shared with `compact-child-cell`, perhaps extract it into a common type?

Reply:
Done.